README.md
# LLM Calculator
This repository contains code to find the optimal runtime to fine-tune the `Llama2-7b` model and optimal codebase.


optimal_llm_calculator/CHANGELOG.md
# Optimal RunTime Calculator

## [0.0.1] - 2023-11-01
Introducing a script which calculates the approximately optimal run time a 
Llama2-7b model will take given some hyperparameters.

### Added
1. Supports all variants of Llama2-7b model.
2. Supports files in `csv`, `parquet`, `json` format.


optimal_llm_codebase/utils.py
"""
File: utils.py
Author: Lokesh Todwal (lokesh.todwal@meesho.com)
"""

from functools import partial

import numpy as np
import pandas as pd
import torch
from datasets import Dataset
from peft import (LoraConfig, TaskType, get_peft_model,
                  prepare_model_for_kbit_training)
from peft.tuners.lora import LoraLayer
from transformers import (LlamaForSequenceClassification, LlamaTokenizerFast,
                          TrainerCallback, TrainerControl, TrainerState,
                          TrainingArguments)

from llama_flash_attn_monkey_patch import replace_llama_attn_with_flash_attn


class SavePeftDeepSpeedModelCallback(TrainerCallback):
    """
    Class to save the deep speed peft model, by assembling the
    states from all the cores. Since we would like to save it
    only for every save_step, we are using modulo concept.
    """

    def __init__(self, trainer, save_steps: int = 3):
        self.trainer = trainer
        self.save_steps = save_steps

    def on_step_end(
        self,
        args: TrainingArguments,
        state: TrainerState,
        control: TrainerControl,
        **kwargs,
    ):
        if state.global_step % self.save_steps == 0:
            self.trainer.accelerator.wait_for_everyone()
            state_dict = self.trainer.accelerator.get_state_dict(self.trainer.deepspeed)
            unwrapped_model = self.trainer.accelerator.unwrap_model(
                self.trainer.deepspeed
            )
            if self.trainer.accelerator.is_main_process:
                unwrapped_model.save_pretrained(args.output_dir, state_dict=state_dict)
            self.trainer.accelerator.wait_for_everyone()
        return control


def tokenize_function(examples, tokenizer):
    """
    Tokenize the textual format data into token_ids and attention masks
    """
    return tokenizer(examples["text"], truncation=True, padding=True, max_length=550)


def create_datasets(tokenizer, use_sfttrainer=False):

    # Repeating this data 3 times to increase the data size
    #   to show how much time it takes
    df = pd.read_csv("optimal_llm_codebase/bitcoin-sentiment-dataset.csv")
    columns = df.columns
    df = pd.DataFrame(np.repeat(df.values, 3, axis=0))
    df.columns = columns
    df.rename(columns={"output": "labels", "input": "text"}, inplace=True)

    text_to_num = {'Positive': 1, 'Negative': 0}
    df['labels'] = df['labels'].apply(lambda x: text_to_num[x])

    train_rto_dataset = Dataset.from_pandas(df.loc[:10000])
    test_rto_dataset = Dataset.from_pandas(df.loc[10000:])

    if use_sfttrainer:
        return train_rto_dataset, test_rto_dataset

    _preprocessing_function = partial(tokenize_function, tokenizer=tokenizer)

    train_tokenized_datasets = train_rto_dataset.map(
        _preprocessing_function, batched=True, remove_columns=["text"]
    )

    test_tokenized_datasets = test_rto_dataset.map(
        _preprocessing_function, batched=True, remove_columns=["text"]
    )

    return train_tokenized_datasets, test_tokenized_datasets


def create_and_prepare_model(args):

    # Flash attention is only supported on A100 or H100 GPU during training
    # due to head dim > 64 backward.
    # Refernce: https://github.com/HazyResearch/flash-attention/issues/190#issuecomment-1523359593 # noqa
    replace_llama_attn_with_flash_attn()

    # NOTE: With DeepSpeed Zero-3, we need to have device_map as None,
    #            else it is incompatible
    model = LlamaForSequenceClassification.from_pretrained(
        args.model_name,
        num_labels=2,
        use_cache=not args.use_gradient_checkpointing,
        quantization_config=None,
        device_map=None,
    )

    # NOTE: Currently Llama is not being supported under SEQ_CLS hence we will
    #   proceed with `https://github.com/huggingface/peft#models-support-matrix` # noqa
    peft_config = LoraConfig(
        lora_alpha=args.lora_alpha,
        lora_dropout=args.lora_dropout,
        r=args.lora_r,
        bias="none",
        task_type=TaskType.SEQ_CLS,
        target_modules=args.lora_target_modules,
    )

    if args.use_gradient_checkpointing:
        model.gradient_checkpointing_enable()

    # Prepare the model for training
    model = prepare_model_for_kbit_training(model)
    model = get_peft_model(model, peft_config)
    peft_module_casting_to_bf16(model, args)
    model.print_trainable_parameters()

    tokenizer = LlamaTokenizerFast.from_pretrained(
        args.model_name, trust_remote_code=True
    )
    tokenizer.pad_token = tokenizer.eos_token
    print("Padding Side: ", tokenizer.padding_side)
    print("Padding Token: ", tokenizer.pad_token)

    return model, tokenizer


def peft_module_casting_to_bf16(model, args):
    for name, module in model.named_modules():
        if isinstance(module, LoraLayer):
            if args.bf16:
                module = module.to(torch.bfloat16)
        if "norm" in name:
            module = module.to(torch.float32)
        if any(x in name for x in ["lm_head", "embed_tokens", "wte", "wpe"]):
            if hasattr(module, "weight"):
                if args.bf16 and module.weight.dtype == torch.float32:
                    module = module.to(torch.bfloat16)


optimal_llm_calculator/optimal_runtime_calculator.py
"""
File: optimal_runtime_calculator.py
Author: Lokesh Todwal (lokesh.todwal@meesho.com)
Description: Script to compute optimal run time for a
              fine-tuning exercise on Llama2-7b model.
"""

import argparse
from builtins import bool, int
from typing import Optional, Tuple

import pandas as pd
from transformers import LlamaTokenizerFast


class OptimalRunTimeCalculator:
    """
    Optimal RunTime Calculator
    """

    def __init__(
        self,
        hf_model_name_or_path: str,
        data_path: str,
        data_field: str,
        number_of_datapoints_to_keep_in_eval: int,
        number_of_datapoints_to_keep_in_training: Optional[int],
        batch_size: int,
        gradient_accumulation_step: int,
        num_of_epochs: int,
        save_steps: int,
        eval_steps: int,
        max_length: int,
        padding: bool,
        truncation: bool,
    ):

        self.hf_model_name_or_path = hf_model_name_or_path
        self.data_path = data_path
        self.data_field = data_field
        self.number_of_datapoints_to_keep_in_eval = number_of_datapoints_to_keep_in_eval
        self.number_of_datapoints_to_keep_in_training = (
            number_of_datapoints_to_keep_in_training
        )
        self.batch_size = batch_size
        self.gradient_accumulation_step = gradient_accumulation_step
        self.num_of_epochs = num_of_epochs
        self.save_steps = save_steps
        self.eval_steps = eval_steps
        self.max_length = max_length
        self.padding = padding
        self.truncation = truncation

        # If number of datapoints to keep in training model is mentioned
        if self.number_of_datapoints_to_keep_in_training:
            _, self.data_sample = self.get_data_metadata_info()
        else:
            # If not, then whole data is used for training the model
            (
                self.number_of_datapoints_to_keep_in_training,
                self.data_sample,
            ) = self.get_data_metadata_info()
        self.tokenized_size = self.get_tokenized_sample_size()

        self.datapoints_in_1_step = self.get_number_of_datapoints_parsed_in_1_step()
        self.number_of_step_in_1_epoch = self.get_number_of_steps_in_1_epoch()

    def get_data_metadata_info(self) -> Optional[Tuple[int, str]]:
        """
        Get metadata information from file
        """

        try:
            if self.data_path.endswith(".csv"):
                df = pd.read_csv(self.data_path)
            elif self.data_path.endswith(".parquet"):
                df = pd.read_parquet(self.data_path)
            elif self.data_path.endswith(".json"):
                df = pd.read_json(self.data_path)

            return df.shape[0], list(df[self.data_field])[0]

        except Exception as exc:
            raise Exception("File path not supported.") from exc

    def get_tokenized_sample_size(self) -> int:
        """
        Get Tokenized data's sample size
        """

        # Tokenizer
        tokenizer = LlamaTokenizerFast.from_pretrained(
            self.hf_model_name_or_path, trust_remote_code=True
        )
        tokenizer.pad_token = tokenizer.eos_token
        return len(
            tokenizer(
                self.data_sample,
                truncation=self.truncation,
                padding=self.padding,
                max_length=self.max_length,
            )["input_ids"]
        )

    def get_number_of_datapoints_parsed_in_1_step(self) -> int:
        """
        Get number of datapoints parsed in 1 step.
        """
        return self.batch_size * self.gradient_accumulation_step * 8

    def get_number_of_steps_in_1_epoch(self) -> int:
        """
        Get number of steps in 1 epoch
        """
        return (
            self.number_of_datapoints_to_keep_in_training // self.datapoints_in_1_step
        )

    def get_iteration_numbers(self) -> Tuple[int, int, int]:
        """
        Get total number of steps, save steps and eval steps.
        """
        total_number_of_steps = (
            self.num_of_epochs * self.get_number_of_steps_in_1_epoch()
        )
        total_number_of_save_steps = total_number_of_steps // self.save_steps
        total_number_of_eval_steps = total_number_of_steps // self.eval_steps

        total_number_of_save_steps = (
            total_number_of_save_steps if total_number_of_save_steps else 1
        )
        total_number_of_eval_steps = (
            total_number_of_eval_steps if total_number_of_eval_steps else 1
        )

        return (
            total_number_of_steps,
            total_number_of_save_steps,
            total_number_of_eval_steps,
        )

    def get_optimized_run_time_for_data(self) -> None:
        """
        Get Optimized RunTime for data.
        """

        (
            total_number_of_steps,
            total_number_of_save_steps,
            total_number_of_eval_steps,
        ) = self.get_iteration_numbers()
        total_number_datapoints_in_1_step = (
            self.get_number_of_datapoints_parsed_in_1_step()
        )

        total_time_taken_by_train_steps = (
            total_number_of_steps
            * 93
            * (self.tokenized_size / 550)
            * (total_number_datapoints_in_1_step / 4096)
        )
        total_time_taken_by_save_steps = total_number_of_save_steps * 23
        total_time_taken_by_eval_steps = (
            total_number_of_eval_steps
            * 44
            * (self.tokenized_size / 550)
            * (self.number_of_datapoints_to_keep_in_eval / 6868)
        )
        print(f"Total number of steps for fine-tuning: "
              f"{total_number_of_steps}.")

        print(f"Total time taken by training steps (approx.): "
              f"{total_time_taken_by_train_steps} seconds.")

        print(f"Total time taken by save steps (approx.): "
              f"{total_time_taken_by_save_steps} seconds.")

        print(f"Total time taken by eval steps (approx.): "
              f"{total_time_taken_by_eval_steps} seconds.")

        total_seconds = int(
            total_time_taken_by_train_steps
            + total_time_taken_by_save_steps
            + total_time_taken_by_eval_steps
        )
        minutes = total_seconds // 60
        secs = total_seconds - minutes * 60
        print(
            f"Total Run time for this experiment will take "
            f"approximately {total_seconds}s (i.e. {minutes}minutes {secs}seconds)"
        )


def main(args):
    optimal_run_time_calculator = OptimalRunTimeCalculator(
        hf_model_name_or_path=args.hf_model_name_or_path,
        data_path=args.data_path,
        data_field=args.data_field,
        number_of_datapoints_to_keep_in_eval=args.number_of_datapoints_to_keep_in_eval,
        number_of_datapoints_to_keep_in_training=args.number_of_datapoints_to_keep_in_training,
        batch_size=args.batch_size,
        gradient_accumulation_step=args.gradient_accumulation_step,
        num_of_epochs=args.num_of_epochs,
        save_steps=args.save_steps,
        eval_steps=args.eval_steps,
        max_length=args.max_length,
        padding=args.padding,
        truncation=args.truncation,
    )

    optimal_run_time_calculator.get_optimized_run_time_for_data()


# ArgParser
parser = argparse.ArgumentParser(description="Optimal Run Time Calculator")
parser.add_argument(
    "--hf_model_name_or_path",
    type=str,
    default="NousResearch/Llama-2-7b-hf",
    help="Variant of Llama2-7b model.",
)
parser.add_argument(
    "--data_path", 
    type=str, 
    required=True, 
    help="Path of the file. Currently supporting CSV, Parquet, JSON."
)
parser.add_argument(
    "--data_field",
    type=str,
    default="text",
    help="Data Field which needs to be tokenized by model.",
)
parser.add_argument(
    "--number_of_datapoints_to_keep_in_eval",
    type=int,
    default=6868,
    help="Size of Evaluation Dataset.",
)
parser.add_argument(
    "--number_of_datapoints_to_keep_in_training",
    type=Optional[int],
    default=None,
    help="Size of Training Data. (If None, the complete data will be used for fine-tuning.",
)
parser.add_argument("--batch_size", type=int, default=8, help="Batch size")
parser.add_argument(
    "--gradient_accumulation_step",
    type=int,
    default=64,
    help="Total number of batch size for which Gradient needs to be accumulated.",
)
parser.add_argument("--num_of_epochs", type=int, default=1, help="Number of epochs.")
parser.add_argument(
    "--save_steps",
    type=int,
    default=3,
    help="After how many steps checkpoint needs to be saved.",
)
parser.add_argument(
    "--eval_steps",
    type=int,
    default=3,
    help="After how many steps do we need to evaluate the validation data.",
)
parser.add_argument(
    "--max_length", type=int, default=1100, help="Maximum length of the token."
)
parser.add_argument(
    "--padding",
    type=bool,
    default=True,
    help="Do we need to pad the tokens to max length if size is less than that?",
)
parser.add_argument(
    "--truncation",
    type=bool,
    default=True,
    help="Truncate the text so as to get the maximum length of the token.",
)

arguments = parser.parse_args()
if __name__ == "__main__":
    main(arguments)


optimal_llm_codebase/llama_flash_attn_monkey_patch.py
# copied from https://github.com/lm-sys/FastChat/blob/main/fastchat/train/llama_flash_attn_monkey_patch.py

from typing import List, Optional, Tuple, Union
import logging

import torch
from torch import nn

import transformers
from transformers.models.llama.modeling_llama import apply_rotary_pos_emb

from einops import rearrange
from flash_attn import flash_attn_func


def forward(
    self,
    hidden_states: torch.Tensor,
    attention_mask: Optional[torch.Tensor] = None,
    position_ids: Optional[torch.LongTensor] = None,
    past_key_value: Optional[Tuple[torch.Tensor]] = None,
    output_attentions: bool = False,
    use_cache: bool = False,
) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
    bsz, q_len, _ = hidden_states.size()

    if self.config.pretraining_tp > 1:
        raise ValueError("pretraining_tp > 1 is not supported for flash attention")
    else:
        query_states = self.q_proj(hidden_states)
        key_states = self.k_proj(hidden_states)
        value_states = self.v_proj(hidden_states)

    query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
    key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
    value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)

    kv_seq_len = key_states.shape[-2]

    if past_key_value is not None:
        kv_seq_len += past_key_value[0].shape[-2]
    cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
    query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)

    if past_key_value is not None:
        # reuse k, v, self_attention
        key_states = torch.cat([past_key_value[0], key_states], dim=2)
        value_states = torch.cat([past_key_value[1], value_states], dim=2)

    past_key_value = (key_states, value_states) if use_cache else None

    query_states, key_states, value_states = [
        rearrange(x, "b h s d -> b s h d") for x in [query_states, key_states, value_states]
    ]

    query_states, key_states, value_states = [x.to(torch.bfloat16) for x in [query_states, key_states, value_states]]
    # print(f"{query.shape=} {key.shape=} {value.shape=}")
    # below output will have shape (batch_size, seqlen, nheads, headdim)
    attn_output = flash_attn_func(query_states, key_states, value_states, causal=True)

    if attn_output.size() != (bsz, q_len, self.num_heads, self.head_dim):
        raise ValueError(
            f"`attn_output` should be of size {(bsz, q_len, self.num_heads, self.head_dim)}, but is"
            f" {attn_output.size()}"
        )

    attn_output = attn_output.reshape(bsz, q_len, self.hidden_size)
    attn_output = self.o_proj(attn_output)
    if output_attentions:
        raise NotImplementedError("`output_attentions` is not supported when `use_flash_attn` is True")
    attn_weights = None

    return attn_output, attn_weights, past_key_value


# Disable the transformation of the attention mask in LlamaModel as the flash attention
# requires the attention mask to be the same as the key_padding_mask
def _prepare_decoder_attention_mask(self, attention_mask, input_shape, inputs_embeds, past_key_values_length):
    # [bsz, seq_len]
    return attention_mask


def replace_llama_attn_with_flash_attn():
    cuda_major, cuda_minor = torch.cuda.get_device_capability()
    if cuda_major < 8:
        logging.warning(
            "Flash attention is only supported on A100 or H100 GPU during training due to head dim > 64 backward."
            "ref: https://github.com/HazyResearch/flash-attention/issues/190#issuecomment-1523359593"
        )
    transformers.models.llama.modeling_llama.LlamaModel._prepare_decoder_attention_mask = (
        _prepare_decoder_attention_mask
    )
    transformers.models.llama.modeling_llama.LlamaAttention.forward = forward


optimal_llm_codebase/train.py
"""
File: train.py
Author: Lokesh Todwal (lokesh.todwal@meesho.com)
Description: Script to fine-tune the LLM model using `deepspeed` and `accelerate`.
"""

import argparse
from typing import List

import datasets
import evaluate
import numpy as np
import pandas as pd
import torch
import transformers
from torch import nn
from transformers import DataCollatorWithPadding, Trainer, TrainingArguments

from utils import (
    SavePeftDeepSpeedModelCallback,
    create_and_prepare_model,
    create_datasets,
)


def compute_metrics(eval_pred):
    """
    Metric to compute along with loss metric for validation dataset.
    """
    # Loading the metric that we want to get along with loss metric
    #     for validation dataset.
    eval_metric_precision = evaluate.load("precision")

    # Note: Adding few popular metrics that one can use for
    #         classification problems.
    # eval_metric_recall = evaluate.load("recall")
    # eval_metric_f1 = evaluate.load("f1")
    # eval_metric_glue_mrpc = evaluate.load("glue", "mrpc")
    # eval_metric_matthews = evaluate.load("matthews_correlation")

    predictions, labels = eval_pred.predictions, eval_pred.label_ids
    predictions = np.argmax(predictions, axis=1)

    precision = eval_metric_precision.compute(
        predictions=predictions, references=labels
    )
    # matthews = eval_metric_matthews.compute(predictions=predictions, references=labels)  # noqa
    # recall = eval_metric_recall.compute(predictions=predictions, references=labels)  # noqa

    # f1 = eval_metric_f1.compute(predictions=predictions, references=labels)
    # glue_mrpc = eval_metric_glue_mrpc.compute(predictions=predictions, references=labels)  # noqa
    # glue_mrpc.update(matthews)

    return precision


def main(arguments):
    """
    Main Method
    """

    class CustomTrainer(Trainer):
        def compute_loss(self, model, inputs, return_outputs=False):
            """
            Custom loss to compute on training and validation dataset.
            """

            with torch.autocast("cuda"):
                labels = inputs.pop("labels")
                # forward pass
                outputs = model(**inputs)
                logits = outputs.get("logits")
                loss_fct = nn.CrossEntropyLoss(
                    weight=torch.tensor([2.35, 0.64], device="cuda")
                )
                loss = loss_fct(
                    logits.view(-1, self.model.config.num_labels),
                    labels.view(-1)
                )

                return (loss, outputs) if return_outputs else loss

    print("Creating Model")
    model, tokenizer = create_and_prepare_model(arguments)
    model.config.use_cache = False

    print("Creating Datasets")
    train_dataset, eval_dataset = create_datasets(tokenizer)

    num_of_steps_in_1_epoch = train_dataset.shape[0] / (
        arguments.per_device_train_batch_size
        * arguments.gradient_accumulation_steps
        * 8
    )

    training_arguments = TrainingArguments(
        output_dir=arguments.output_dir,
        save_strategy="steps",
        evaluation_strategy="steps",
        logging_strategy="steps",
        eval_steps=arguments.eval_steps,
        save_steps=arguments.save_steps,
        logging_steps=arguments.eval_steps,
        max_steps=int(num_of_steps_in_1_epoch),
        per_device_train_batch_size=arguments.per_device_train_batch_size,
        auto_find_batch_size=True,
        per_device_eval_batch_size=arguments.per_device_eval_batch_size,
        gradient_accumulation_steps=arguments.gradient_accumulation_steps,
        optim=arguments.optim,
        learning_rate=5e-5,
        bf16=arguments.bf16,
        max_grad_norm=arguments.max_grad_norm,
        warmup_ratio=arguments.warmup_ratio,
        lr_scheduler_type=arguments.lr_scheduler_type,
        gradient_checkpointing=arguments.use_gradient_checkpointing,
        report_to=["wandb"],
        save_total_limit=1,
        label_names=arguments.label_name,
        disable_tqdm=False,
        metric_for_best_model=arguments.metric_for_best_model,
        greater_is_better=arguments.greater_is_better,
        load_best_model_at_end=True,
    )

    # Not passing `padding` and `max_length` DataCollatorWithPadding as I have already  # noqa
    #     created constant length tokenized input_ids and attention_mask.
    trainer = CustomTrainer(
        model=model,
        args=training_arguments,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
        tokenizer=tokenizer,
        data_collator=DataCollatorWithPadding(tokenizer),
        compute_metrics=compute_metrics,
    )
    # trainer.accelerator.print(f"{trainer.model}")

    trainer.accelerator.print(
        "Number of steps in 1 epochs: ",
        num_of_steps_in_1_epoch
    )

    trainer.accelerator.print(
        "Number of datapoints parsed in 1 step: ",
        arguments.per_device_train_batch_size
        * arguments.gradient_accumulation_steps
        * 8,
    )

    if trainer.accelerator.is_main_process:
        datasets.utils.logging.set_verbosity_warning()
        transformers.utils.logging.set_verbosity_info()
    else:
        datasets.utils.logging.set_verbosity_error()
        transformers.utils.logging.set_verbosity_error()

    trainer.accelerator.print("Adding DeepSpeed PEFT Callback.")
    trainer.add_callback(
        SavePeftDeepSpeedModelCallback(trainer, save_steps=arguments.save_steps)
    )

    trainer.accelerator.print("Starting Training")
    train_result = trainer.train()
    trainer.accelerator.print("Metrics: ", train_result.metrics)

    trainer.accelerator.print("Saving Final Model")
    trainer.accelerator.print("Saving pretrained model using DeepSpeed PEFT.")
    trainer.accelerator.wait_for_everyone()
    state_dict = trainer.accelerator.get_state_dict(trainer.deepspeed)
    unwrapped_model = trainer.accelerator.unwrap_model(trainer.deepspeed)
    if trainer.accelerator.is_main_process:
        unwrapped_model.save_pretrained(
            arguments.output_dir,
            state_dict=state_dict
        )
    trainer.accelerator.wait_for_everyone()

    df = pd.DataFrame(trainer.state.log_history)
    df.to_csv("./log_history.csv", index=False)


parser = argparse.ArgumentParser(description="Codebase to fine-tune model optimally.")
parser.add_argument(
    "--model_name",
    type=str,
    default="NousResearch/Llama-2-7b-hf",
    help="LLM HuggingFace model or path.",
)
parser.add_argument(
    "--lora_r", type=int, default=16, help="Rank of LoRA decomposed matrix."
)
parser.add_argument(
    "--lora_alpha",
    type=int,
    default=64,
    help="LoRA Scaling factor.",
)
parser.add_argument(
    "--lora_dropout",
    type=float,
    default=0.1,
    help="LoRA dropout.",
)
parser.add_argument(
    "--lora_target_modules",
    type=List[str],
    default=["self_attn.q_proj", "self_attn.v_proj"],
    help="Projection of weights on which matrix needs to be decomposed.",
)
parser.add_argument(
    "--per_device_train_batch_size",
    type=int,
    default=8,
    help="The batch size per GPU/XPU/TPU/MPS/NPU core/CPU for training.",
)
parser.add_argument(
    "--per_device_eval_batch_size",
    type=int,
    default=64,
    help="The batch size per GPU/XPU/TPU/MPS/NPU core/CPU for evaluation.",
)
parser.add_argument(
    "--gradient_accumulation_steps",
    type=int,
    default=64,
    help="Number of updates steps to accumulate the gradients for, "
         "before performing a backward/update pass.",
)
parser.add_argument(
    "--optim", type=str, default="adafactor", help=" The optimizer to use."
)
parser.add_argument(
    "--max_grad_norm",
    type=float,
    default=0.9,
    help="Maximum gradient norm (for gradient clipping)",
)
parser.add_argument(
    "--warmup_ratio",
    type=float,
    default=0.03,
    help="Ratio of total training steps used for a linear "
         "warmup from 0 to learning_rate.",
)
parser.add_argument(
    "--lr_scheduler_type",
    type=str,
    default="linear",
    help="The scheduler type to use."
)
parser.add_argument(
    "--label_name",
    type=List[str],
    default=["labels"],
    help="The name of the target label.",
)
parser.add_argument(
    "--num_train_epochs",
    type=str,
    default="linear",
    help="The scheduler type to use."
)
parser.add_argument(
    "--use_gradient_checkpointing",
    type=bool,
    default=True,
    help="Gradient checkpointing to save memory at the expense of slower backward pass.",
)
parser.add_argument(
    "--bf16",
    type=bool,
    default=True,
    help="The scheduler type to use."
)
parser.add_argument(
    "--output_dir",
    type=str,
    default="output",
    help="The output directory where the model predictions "
         "and checkpoints will be written.",
)
parser.add_argument(
    "--save_steps",
    type=int,
    default=3,
    help="After how many steps checkpoint needs to be saved.",
)
parser.add_argument(
    "--eval_steps",
    type=int,
    default=3,
    help="After how many steps do we need to evaluate the validation data.",
)
parser.add_argument(
    "--max_length", type=int, default=1100, help="Maximum length of the token."
)
parser.add_argument(
    "--metric_for_best_model",
    type=str,
    default="loss",
    help="Metric to use to compare two different models.",
)
parser.add_argument(
    "--greater_is_better",
    type=bool,
    default=False,
    help="Specify if better models should have a greater metric or not",
)

args = parser.parse_args()

if __name__ == "__main__":
    main(args)
